lms/djangoapps/courseware/views/views.py

```diff
@@ -749,42 +739,101 @@ def _progress(request, course_key, student_id):
+            missing_required_verification = enrollment_mode in CourseMode.VERIFIED_MODES and \
+                not SoftwareSecurePhotoVerification.user_is_verified(student)
+            if cert_downloadable_status['is_downloadable']:
+                cert_status = 'downloadable'
+                title = 'Your certificate is available'
+                msg = 'You can keep working for a higher grade, or request your certificate now.'
+                download_url = cert_downloadable_status['download_url']
+            elif cert_downloadable_status['is_generating']:
+                cert_status = 'generating'
+                title = "We're working on it..."
+                msg = "We're creating your certificate. You can keep working in your courses and a link to it will " \
+                      "appear here and on your Dashboard when it is ready."
+            elif missing_required_verification or cert_downloadable_status['is_unverified']:
+                platform_name = theming_helpers.get_value('PLATFORM_NAME', settings.PLATFORM_NAME)
+                cert_status = 'unverified'
+                title = 'Certificate unavailable'
+                msg = 'You have not received a certificate because you do not have a current {platform_name} ' \
+                      'verified identity.'.format(platform_name=platform_name)
+            else:
+                cert_status = 'request_cert'
+                title = 'Congratulations, you qualified for a certificate!'
+                msg = 'You can keep working for a higher grade, or request your certificate now.'
+
+            if certs_api.has_html_certificates_enabled(course_key, course):
+                if certs_api.get_active_web_certificate(course) is not None:
+                    show_cert_web_view = True
+                    cert_web_view_url = certs_api.get_certificate_url(
+                        course_id=course_key, uuid=cert_downloadable_status['uuid']
+                    )
+                else:
+                    download_url = None
+                    cert_status = 'generating'
+                    title = "We're working on it..."
+                    msg = "We're creating your certificate. You can keep working in your courses and a link to it " \
+                          "will appear here and on your Dashboard when it is ready."
+    elif enrollment_mode == CourseMode.AUDIT:
+        cert_status = 'not-applicable'
+        title = 'Your enrollment: Audit track'
+        msg = 'You are enrolled in the audit track for this course. The audit track does not include a certificate.'
+    else:
+        return None
+
+    response_data = {
+        'cert_status': cert_status,
+        'title': _(title),  # pylint: disable=translation-of-non-string
+        'msg': _(msg),  # pylint: disable=translation-of-non-string
+        'download_url': download_url,
+        'show_cert_web_view': show_cert_web_view,
+        'cert_web_view_url': cert_web_view_url
+    }
```

Comment:
It would be better to make this an object (either by making a class, or by at least making a `namedtuple`). Passing around dicts with constant keys often turns into a maintenance headache down the road.